_test/test_sqw_class/test_equal_to_tol.m

```diff
@@ -0,0 +1,270 @@
+        times_taken_partial = benchmark_function(f_partial, num_iters, num_reps);
+        median_time_partial = median(times_taken_partial);
+
+        assertTrue(median_time_full > median_time_partial);
+    end
+
+    function test_equal_to_tol_true_for_eq_sqw_reorder_false_with_fraction(obj)
+        sqw_copy = copy(obj.sqw_2d_paged);
+        assertTrue( ...
+            equal_to_tol( ...
+                obj.sqw_2d_paged, ...
+                sqw_copy, ...
+                'fraction', 0.5, ...
+                'reorder', false ...
+            ) ...
+        );
+    end
+
+    function test_equal_to_tol_can_be_called_with_negative_tol_for_rel_tol(obj)
+        sqw_copy = copy(obj.sqw_2d_paged);
+        rel_tol = 1e-5;
+        assertTrue(equal_to_tol(sqw_copy, obj.sqw_2d_paged, -rel_tol));
+
+        % find first non-zero signal value
+        sig_idx = find(sqw_copy.data.s > 0, 1);
+
+        % increase the difference in the value by 1% more than rel_tol
+        value_diff = 1.01*rel_tol*obj.sqw_2d_paged.data.s(sig_idx);
+        sqw_copy.data.s(sig_idx) = obj.sqw_2d_paged.data.s(sig_idx) + value_diff;
+
+        assertFalse(equal_to_tol(sqw_copy, obj.sqw_2d_paged, -rel_tol));
+
+        % check increasing the rel_tol by 1% returns true
+        assertTrue(equal_to_tol(sqw_copy, obj.sqw_2d_paged, -rel_tol*1.01))
+
+        % check absolute tolerance still true
+        assertTrue(equal_to_tol(sqw_copy, obj.sqw_2d_paged, value_diff + 1e-8));
+    end
+
+    function test_objects_are_not_equal_if_one_has_empty_pixeldata(obj)
+        non_empty_sqw = obj.sqw_2d;
+        empty_sqw = obj.sqw_2d;
+        empty_sqw.data.pix = PixelData();
+
+        assertFalse(equal_to_tol(empty_sqw, non_empty_sqw));
+        assertFalse(equal_to_tol(non_empty_sqw, empty_sqw));
+    end
+end
+
+methods (Static)
+
+    function shuffled_pix = shuffle_pixel_bin_rows(pix, npix)
+        % Shuffle the pixels in the bins defined by the npix array
+        %
+        npix_non_empty = npix(npix ~= 0);
+        shuffled_pix = PixelData(pix.num_pixels);
+
+        bin_end_idxs = cumsum(npix_non_empty(:));
+        bin_start_idxs = bin_end_idxs - npix_non_empty(:) + 1;
+        for i = 1:numel(npix_non_empty)
+            bin_start = bin_start_idxs(i);
+            bin_end = bin_end_idxs(i);
+
+            pix_in_bin = pix.data(:, bin_start:bin_end);
+            shuffled_bin_pix = pix_in_bin(:, randperm(size(pix_in_bin, 2)));
+            shuffled_pix.data(:, bin_start:bin_end) = shuffled_bin_pix;
+
+            % Checksum to ensure the bins' columns contain the same values
+            assertEqual(sum(shuffled_bin_pix, 2), sum(pix_in_bin, 2), '', 1e-8);
+        end
+    end
+
+    function pix = get_pix_with_fake_faccess(data, npix_in_page)
+        faccess = FakeFAccess(data);
+        bytes_in_pixel = PixelData.DATA_POINT_SIZE*PixelData.DEFAULT_NUM_PIX_FIELDS;
+        pix = PixelData(faccess, npix_in_page*bytes_in_pixel);
+    end
+
+end
+
+end
```

Comment:
I have not looked at much of the tests in the interests of a rapid review

Reply:
This file has been copied largely unchanged - the only addition is a new `test_objects_are_not_equal_if_one_has_empty_pixeldata`) from `master`